packages/tpetra/core/example/BlockCrs/Tpetra_TestBlockCrs.cpp

```diff
@@ -124,537 +124,541 @@ int main (int argc, char *argv[])
-
-             LO i0, j0, k0, i1, j1, k1;
-             sb.idx_to_ijk(gid_row, i0, j0, k0);
-             sb.idx_to_ijk(gid_col, i1, j1, k1);
-
-             const LO diff_i = i0 - i1;
-             const LO diff_j = j0 - j1;
-             const LO diff_k = j0 - k1;
-
-             auto block = blocks(loc);
-             for (LO l0=0;l0<blocksize;++l0)
-               for (LO l1=0;l1<blocksize;++l1)
-                 block(l0, l1) = get_block_crs_entry<value_type>(i0, j0, k0,
-                                                                 diff_i, diff_j, diff_k,
-                                                                 l0, l1);
-           }
-        });
-    }
+      if (verbose) {
+        bcrs_graph->describe(*out, Teuchos::VERB_EXTREME);
+      }
 
-    {
-      TimeMonitor timerBlockCrsFillComplete(*TimeMonitor::getNewTimer("3) BlockCrsMatrix FillComplete - currently do nothing"));
-      // this function does not exist
-      //A_bcrs->fillComplete();
-    }
+      if (debug) {
+        std::ostringstream os;
+        os << *debugPrefix << "Make BlockCrsMatrix" << endl;
+        std::cerr << os.str ();
+      }
+      // Create BlockCrsMatrix
+      RCP<tpetra_blockcrs_matrix_type> A_bcrs (new tpetra_blockcrs_matrix_type (*bcrs_graph, blocksize));
 
-    if (verbose) {
-      A_bcrs->describe(*out, Teuchos::VERB_EXTREME);
-    }
+      if (debug) {
+        std::ostringstream os;
+        os << *debugPrefix << "Fill BlockCrsMatrix" << endl;
+        std::cerr << os.str ();
+      }
+      typedef tpetra_blockcrs_matrix_type::little_block_type block_type;
+      Kokkos::View<block_type*,exec_space> blocks;
+      {
+        TimeMonitor timerLocalBlockCrsFill(*TimeMonitor::getNewTimer("2) LocalBlockCrsFill"));
+
+        // Tpetra BlockCrsMatrix only has high level access functions
+        // To fill this on device, we need an access to the meta data of blocks
+        const auto rowptr_host = Kokkos::create_mirror_view(rowptr);
+        const auto colidx_host = Kokkos::create_mirror_view(colidx);
+
+        Kokkos::deep_copy(rowptr_host, rowptr);
+        Kokkos::deep_copy(colidx_host, colidx);
+
+        blocks = Kokkos::View<block_type*,exec_space>("blocks", rowptr_host(num_owned_elements));
+
+        const auto blocks_host = Kokkos::create_mirror_view(blocks);
+        // This MUST run on host, since it invokes a host-only method,
+        // getLocalBlock.  This means we must NOT use KOKKOS_LAMBDA,
+        // since that would build the lambda for both host AND device.
+        Kokkos::parallel_for
+          (Kokkos::RangePolicy<host_space, LO> (0, num_owned_elements),
+           [=] (const LO row) {
+            const auto beg = rowptr_host(row);
+            const auto end = rowptr_host(row+1);
+            typedef typename std::remove_const<decltype (beg) >::type offset_type;
+            for (offset_type loc = beg; loc < end; ++loc) {
+              blocks_host(loc) = A_bcrs->getLocalBlock(row, colidx(loc));
```

Comment:
I'm not sure if it's OK to capture RCP by value in a thread-parallel loop, even though it's on host.  I would recommend capturing `const BlockCrsMatrix&` instead.

Reply:
umm... you are right. it should be captured by reference.

Reply:
@kyungjoo-kim This is the only part I'm really worried about, and it's only a big deal if the host execution space is OpenMP.  It's unwise to take RCP inside a parallel loop.

Reply:
Now I have some curiociy. If using rcp is not a good idea in a parallel loop, why is kokkos view reference counting okay ? For the same reason, block_host will increase the view reference count multiple times. Does it use atomic ? Does this mean that using unmanaged views inside of a parallel loop is always better than using using managed views ? KOKKOS_LAMBDA basically capture by value. On GPUs, does this cause the kokkos reference count to explode ? Just curiocity.

Reply:
> If using rcp is not a good idea in a parallel loop, why is kokkos view reference counting okay ?

1. Kokkos switches off reference counting for Views inside parallel regions.
2. RCP's reference count is not thread safe by default.  (There's a nondefault build option to enable this, but it does have a run-time cost.)  `std::shared_ptr`'s reference count _is_ thread safe with respect to `std::thread`.

Reply:
> On GPUs, does this cause the kokkos reference count to explode ?

RCP's methods don't have device function markings, so they should not even compile in a device kernel.  Even if they did have device function markings, RCP's reference count lives in a pointer to host memory (a struct that has the reference count in it), so it's not accessible on device if you created the RCP on host.

Reply:
This should actually be fine -- @kyungjoo-kim captures the RCP by reference.  Better practice would be to capture a reference to the object, but this is not incorrect.